controller/mutators/etcd/membership_mutator.go

```diff
@@ -0,0 +1,127 @@
+//  Copyright (c) 2017-2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+import (
+	"errors"
+	"strconv"
+	"unsafe"
+
+	"github.com/m3db/m3/src/cluster/services"
+	"github.com/uber/aresdb/cluster/kvstore"
+	"github.com/uber/aresdb/controller/models"
+	"github.com/uber/aresdb/controller/mutators/common"
+	"github.com/uber/aresdb/utils"
+)
+
+// NewMembershipMutator creates new MembershipMutator
+func NewMembershipMutator(etcdClient *kvstore.EtcdClient) common.MembershipMutator {
+	return membershipMutatorImpl{
+		etcdClient: etcdClient,
+	}
+}
+
+type membershipMutatorImpl struct {
+	etcdClient *kvstore.EtcdClient
+}
+
+func (mm membershipMutatorImpl) Join(namespace string, instance models.Instance) error {
+	return errors.New("join not supported in etcd version of membership mutator")
```

Comment:
we depend on datanode heart beat instead?

Reply:
yeah. easiest route i think

Reply:
it is not through controller, each datanode heartbeat using m3db's heartbeat service

Reply:
let's have some talk to make it clear